File: common/compute/ebs/ebs_factory.py
from aws_cdk import aws_ec2 as ec2, Size
from constructs import Construct
from abc import ABC, abstractmethod


class EBSStorage(ABC):
    @abstractmethod
    def initialize(self, scope: Construct, vpc: ec2.Vpc):
        pass


class GeneralPurposeEbs(EBSStorage):
    def __init__(self, value: str):
        super().__init__()
        print(value)

    def initialize(self, scope: Construct, vpc: ec2.Vpc):
        # General Purpose SSD (gp2)
        ec2.Volume(
            scope,
            "MyEbsVolume",
            availability_zone=vpc.availability_zones[0],
            size=Size.gibibytes(10),
            volume_type=ec2.EbsDeviceVolumeType.GP2,
        )


class IopsEbs(EBSStorage):
    def __init__(self, value: str):
        super().__init__()
        print(value)

    def initialize(self, scope: Construct, vpc: ec2.Vpc):
        # Provisioned IOPS SSD (io1)
        ec2.Volume(
            scope,
            "MyIopsEbsVolume",
            availability_zone=vpc.availability_zones[0],
            size=Size.gibibytes(20),
            volume_type=ec2.EbsDeviceVolumeType.IO1,
            iops=1000,  # Set IOPS for IO1 volumes
        )


class ColdHddEbs(EBSStorage):
    def __init__(self, value: str):
        super().__init__()
        print(value)

    def initialize(self, scope: Construct, vpc: ec2.Vpc):
        # Cold HDD (sc1)
        ec2.Volume(
            scope,
            "MyColdHddEbsVolume",
            availability_zone=vpc.availability_zones[0],
            size=Size.gibibytes(500),
            volume_type=ec2.EbsDeviceVolumeType.SC1,
        )


class AbstractFactory(ABC):
    @abstractmethod
    def create_ebs(self, volume_type: str) -> EBSStorage:
        pass


class EbsFactory(AbstractFactory):
    def create_ebs(self, volume_type: str) -> EBSStorage:
        if volume_type == "gp2":
            return GeneralPurposeEbs("General Purpose EBS Created.")
        elif volume_type == "io1":
            return IopsEbs("Provisioned IOPS EBS Created.")
        elif volume_type == "sc1":
            return ColdHddEbs("Cold HDD EBS Created.")
        else:
            raise ValueError(f"Unknown volume type: {volume_type}")

File: csharp/src/Csharp/common/ebsFactory.cs
using Amazon.CDK;
using Amazon.CDK.AWS.EC2;
using Constructs;
using System;

namespace Csharp.common
{
    // Abstract base class for EBS storage
    public abstract class EbsStorage
    {
        public abstract void Initialize(Construct scope, Vpc vpc);
    }

    // General Purpose EBS (gp2) implementation
    public class GeneralPurposeEbs : EbsStorage
    {
        public GeneralPurposeEbs(string value)
        {
            Console.WriteLine(value);
        }

        public override void Initialize(Construct scope, Vpc vpc)
        {
            new Volume(scope, "MyEbsVolume", new VolumeProps
            {
                AvailabilityZone = vpc.AvailabilityZones[0],
                Size = Size.Gibibytes(10),
                VolumeType = EbsDeviceVolumeType.GP2
            });
        }
    }

    // Provisioned IOPS EBS (io1) implementation
    public class IopsEbs : EbsStorage
    {
        public IopsEbs(string value)
        {
            Console.WriteLine(value);
        }

        public override void Initialize(Construct scope, Vpc vpc)
        {
            new Volume(scope, "MyIopsEbsVolume", new VolumeProps
            {
                AvailabilityZone = vpc.AvailabilityZones[0],
                Size = Size.Gibibytes(20),
                VolumeType = EbsDeviceVolumeType.IO1,
                Iops = 1000  // Set IOPS for IO1 volumes
            });
        }
    }

    // Cold HDD EBS (sc1) implementation
    public class ColdHddEbs : EbsStorage
    {
        public ColdHddEbs(string value)
        {
            Console.WriteLine(value);
        }

        public override void Initialize(Construct scope, Vpc vpc)
        {
            new Volume(scope, "MyColdHddEbsVolume", new VolumeProps
            {
                AvailabilityZone = vpc.AvailabilityZones[0],
                Size = Size.Gibibytes(500),
                VolumeType = EbsDeviceVolumeType.SC1
            });
        }
    }

    // Abstract Factory for creating EBS storage
    public abstract class AbstractFactory
    {
        public abstract EbsStorage CreateEbs(string volumeType);
    }

    // Concrete EBS Factory implementation
    public class EbsFactory : AbstractFactory
    {
        public override EbsStorage CreateEbs(string volumeType)
        {
            switch (volumeType)
            {
                case "gp2":
                    return new GeneralPurposeEbs("General Purpose EBS Created.");
                case "io1":
                    return new IopsEbs("Provisioned IOPS EBS Created.");
                case "sc1":
                    return new ColdHddEbs("Cold HDD EBS Created.");
                default:
                    throw new ArgumentException($"Unknown volume type: {volumeType}");
            }
        }
    }
}
